### PR TITLE
MVP Benthos Validator

### PR DESCRIPTION
**Pull request by geowurster**
Here's what I have so far to address https://github.com/SkyTruth/Benthos/issues/212.

I put this in utils to keep it out of the way but it will likely need to go somewhere importable and be given a better name.  I modified the message validator function structure a little bit to include a good and bad value so I could include a little unittest at the end - to run it do: `$ ./utils/benthos_validate.py test`.  For every type we're checking it builds a good message, tests it, and then builds bad messages by changing one field at a time.

In order to incorporate this into the test suite we need to do something like this:

```python
import json
import sys

import benthos_validate

def newline_parser(infile):
    for line in infile:
        yield json.loads(line)

with open('Messages-To-Validate.json') as f:
    assert benthos_validate.validate_messages(newline_parser(f), err=sys.stdout)
```

Right now messages are evaluated in bulk and a single `True/False` value is returned.  Information about why any given field/message field is written to a file if it is supplied.

Additional notes can be found on line 15 in `utils/benthos_validate.py`.

### CODE REVIEW

**Comment by geowurster** [comment]
@redhog I should have done this on a branch.  Let me know if this causes a problem with anything you're working on and I'll move it to one.

**Comment by redhog** [comment]
@geowurster is there a reason this code doesn't use / incorporate https://github.com/SkyTruth/Benthos/blob/master/benthos/_schema_def.py and https://github.com/SkyTruth/Benthos/blob/master/benthos/schema.py ?

**Comment by geowurster** [comment]
I thought about doing that but didn't want to make installing all of Benthos a dependency and the schema just defines the expected type without considering acceptable values.  For example the schema doesn't care if the field `epfd` is -1 but the validator checks to make sure the value is in its acceptable range of 0-9.

**Comment by redhog** [comment]
I mean that we should migrate that code into libais, removing it from benthos altogether...

**Comment by redhog** [comment]
And it should definitely be extended by range checks :)

**Comment by geowurster** [comment]
Well, here's the MVP!

**Comment by redhog** [comment]
About moving code from benthos to pelagos etc - it would be much better if all the format handling (gpsd with msgpack container, so benthos.schema and benthos.io.gpsd) was done in a non-pelagos library, and I was thinking that the python library part of libais could work for that...
what are your thoughts?

File: ais/stream/__init__.py [hunk 1]
```diff
@@ -116,7 +154,7 @@ def normalize(nmea=sys.stdin,
           station = fields[i]
           break  # Found it so ditch the for loop.
 
-      if station is None and options.allowUnknown:
+      if station is None and allowUnknown:
```
Comment (schwehr) [inline ais/stream/__init__.py]: Would prefer:

```
if not station and allowUnknown
```

File: utils/benthos_validate.py [hunk 2]
```diff
@@ -0,0 +1,406 @@
+#!/usr/bin/env python
+
+
+"""
+Validator to ensure libais can produce Benthos's expected input
+"""
```
Comment (schwehr) [inline utils/benthos_validate.py]: Missing a period at the end of the summary sentence.

Best to go with one big doc string.  The comments below should go into the doc str.

File: utils/benthos_validate.py [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+            for field in msg_fields:
+                if field != 'type':
+                    bad_message = good_message.copy()
+                    bad_message[field] = MSG_VALIDATION_LOOKUP[field]['bad']
+                    assert not validate_messages([bad_message]), \
+                        "Field `%s' should have caused message to fail: %s" % (field, bad_message)
+
+
+if __name__ == '__main__':
+    if len(sys.argv) > 1 and sys.argv[1].lower() == 'test':
```
Comment (schwehr) [inline utils/benthos_validate.py]: why not just this?

```
if __name__ == '__main__':
    unittest.main()

**Comment by redhog** [comment]
I've started to move all benthos validation code over to https://github.com/SkyTruth/gpsd_format :)

**Comment by geowurster** [comment]
Closing in favor of [gpsdio](https://github.com/skytruth/gpsd_format)

**Comment by geowurster** [comment]
@schwehr I wasn't sure how `unittest.main()` handled exiting and thought it might just return the desired exit code.  I just checked it and it handles exiting on its own.

**Pull request closed without merging**